src/include/types.h

```diff
@@ -541,5 +541,33 @@ WRITE_CLASS_ENCODER(errorcode32_t)
 WRITE_EQ_OPERATORS_1(errorcode32_t, code)
 WRITE_CMP_OPERATORS_1(errorcode32_t, code)
 
+struct sha1_fp_t {
+#define SHA1_DIGEST_SIZE 20
+  unsigned char v[SHA1_DIGEST_SIZE];
+  bool valid;
```

Comment:
I'm not sure that the bool valid is necessary.  The default ctor can just zero the sha1, and if we really need to represent an undefined value we can use boost::optional<sha1_digest_t>?